auth/auth/auth.py

```diff
@@ -0,0 +1,233 @@
+    scopes = [
+        'https://www.googleapis.com/auth/userinfo.profile',
+        'https://www.googleapis.com/auth/userinfo.email',
+        'openid'
+    ]
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        '/auth-oauth2-client-secret/client_secret.json', scopes=scopes, state=state)
+    flow.redirect_uri = redirect_uri
+    return flow
+
+
+@routes.get('/healthcheck')
+async def get_healthcheck(request):  # pylint: disable=W0613
+    return web.Response()
+
+
+@routes.get('')
+@routes.get('/')
+async def get_index(request):  # pylint: disable=unused-argument
+    return aiohttp.web.HTTPFound('/login')
+
+
+@routes.get('/login')
+@web_maybe_authenticated_user
+async def login(request, userdata):
+    next = request.query.get('next', deploy_config.external_url('notebook2', ''))
+    if userdata:
+        return aiohttp.web.HTTPFound(next)
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'))
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        include_granted_scopes='true')
+
+    session = await aiohttp_session.get_session(request)
+    session['state'] = state
+    session['next'] = next
+
+    return aiohttp.web.HTTPFound(authorization_url)
+
+
+@routes.get('/oauth2callback')
+async def callback(request):
+    session = await aiohttp_session.get_session(request)
+    if 'state' not in session:
+        raise web.HTTPUnauthorized()
+    state = session['state']
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'), state=state)
+
+    try:
+        flow.fetch_token(code=request.query['code'])
+        token = google.oauth2.id_token.verify_oauth2_token(
+            flow.credentials.id_token, google.auth.transport.requests.Request())
+        id = token['sub']
+    except Exception:
+        log.exception('oauth2 callback: could not fetch and verify token')
+        raise web.HTTPUnauthorized()
+
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('SELECT * from user_data where user_id = %s;', f'google-oauth2|{id}')
+            users = await cursor.fetchall()
+
+    if len(users) != 1:
+        raise web.HTTPUnauthorized()
+    user = users[0]
+
+    session_id = await create_session(dbpool, user['id'])
+
+    del session['state']
+    session['session_id'] = session_id
+    next = session.pop('next')
+    return aiohttp.web.HTTPFound(next)
+
+
+@routes.post('/logout')
+@web_maybe_authenticated_user
+async def logout(request, userdata):
+    if not userdata:
+        return web.HTTPFound(deploy_config.external_url('notebook2', ''))
+
+    dbpool = request.app['dbpool']
+    session_id = userdata['session_id']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('DELETE FROM sessions WHERE session_id = %s;', session_id)
```

Comment:
We may want to do something besides deleting sessions.

I may be interested in knowing how many sessions a user has had, distribution of session creation times, approximate geographic locations.

What do you think about setting a boolean (valid) flag on the session, or just relying on logout (more normalized): (login: Date, logout: Date, ip: ).

Then on login we can control whether we want to grant the user a session, or alert them to some malarkey.

Reply:
Yes, great idea.  I think there are ultimately going to be compliance requirements that track logs (as well as other things, like access) so I think we should revisit this question when those requirements are more clear.